### PR TITLE
feat(attention): add SM120 sparse MLA kernels

### PR DESCRIPTION
**Pull request by lucifer1004**
## Summary

Sparse-MLA paged attention for SM120a / SM121a, covering both the DSv4 family
(`d_qk=512`) and the DSv3.2 / GLM-5.1 family (`d_qk=576`). Adds decode
(warp-spec, split-K) and prefill (TMA + warp-spec) kernels in the SM120 sparse
MLA JIT/AOT module, with dispatch through the existing MLA Python APIs.

## Surfaces

The user-facing surface is limited to extending existing MLA entry points with
SM120/SM121 sparse backend coverage. SM120-specific dispatchers, runners, and
direct kernel wrappers remain internal implementation details; users should
continue to call the public `flashinfer.mla` APIs below. The public APIs keep
their existing decode-oriented names, while the SM120 sparse backend internally
selects decode or prefill kernels from the flattened token count.

### Public Python APIs

- `flashinfer.mla.trtllm_batch_decode_sparse_mla_dsv4(...)`
  - Extends the DSv4 sparse-MLA decode entry point to SM120/SM121.
  - SM100/SM103 continue to route through the TRTLLM-GEN path.
  - SM120/SM121 route through the new packed sparse backend. The SM120 path
    accepts BF16 query tensors, a packed uint8 SWA KV cache with 584 bytes per
    token, optional attention sinks, and an optional second packed KV cache via
    `extra_sparse_indices` / `extra_sparse_topk_lens`.
  - `swa_topk_lens` describes the active SWA segment. When the optional
    compressed segment is present, `extra_sparse_topk_lens` describes its active
    length. Both sparse segments share one online-softmax denominator.

- `flashinfer.mla.trtllm_batch_decode_with_kv_cache_mla(...)`
  - Extends the existing MLA decode API with an SM120/SM121 sparse backend for
    DSv3.2 / GLM-style packed sparse MLA (`d_qk=576`, 656 bytes per token).
  - With `backend="auto"` on SM120/SM121 and `sparse_mla_top_k > 0`, the call
    selects the sparse backend. `backend="sparse"` selects it explicitly.
  - `kv_scale_format` controls the packed-cache scale convention:
    `"auto"` / `"pow2_fp32"` for DSv3.2-style power-of-two FP32 scales, and
    `"arbitrary_fp32"` for GLM-style arbitrary FP32 scales.

### Internal / JIT Implementation Details

- `flashinfer.mla._sparse_mla_sm120._sparse_mla_sm120_paged_attention(...)` is
  an internal dispatcher that shares validation and buffer handling between
  decode and prefill shapes. It is registered as the custom op
  `flashinfer::sparse_mla_sm120_paged_attention`.
- `_SparseMLAPagedAttentionRunner` is an internal helper used by tests,
  benchmarks, and the public API implementation to own LSE and decode scratch
  buffers.
- `flashinfer.mla._sparse_mla_sm120.sparse_mla_sm120_decode_dsv4(...)` and the
  DSv3.2 decode wrapper are internal direct kernel wrappers around the JIT
  module and AutoTuner `chunks_per_block` tactic cache.
- `flashinfer.jit.mla.gen_sparse_mla_sm120_module()` wires the new SM120 sparse
  MLA CUDA sources into the JIT/AOT build.

## Performance

RTX PRO 6000 Blackwell Server Edition (SM120a). KV pool sized > L2 (>=600 MB)
so analytical numbers reflect DRAM-bound production. `attn_TFLOPs = 2 * T *
num_heads * topk * (d_qk + d_v) / latency`.

**DSv4 prefill, single cache** (584 B/slot footer, FP8 paged):

| NH | topk | T | latency us | KV GB/s | TFLOPs |
|---:|---:|---:|---:|---:|---:|
|  16 |  128 |  256 |    53.2 | 359 |  20.2 |
|  64 |  128 |  256 |   108.5 | 176 |  39.6 |
|  64 | 2048 |  256 |   608.3 | 503 | 113.0 |
| 128 | 2048 |  256 |  1171.5 | 261 | 117.3 |
|  64 | 2048 | 1024 |  2096.1 | 584 | 131.1 |
| 128 | 2048 | 1024 |  4110.4 | 298 | 133.8 |

**DSv4 prefill, dual cache** (main SWA cache `topk=128` plus secondary C4A /
C128A cache, 584 B/slot footer, FP8 paged):

| NH | main topk | extra topk | T | extra PBS | latency us | KV GB/s | TFLOPs |
|---:|---:|---:|---:|---:|---:|---:|---:|
|  64 | 128 |  512 | 256 | 64 |   239.6 | 399 |  89.6 |
| 128 | 128 |  512 | 512 | 64 |   804.8 | 238 | 106.7 |
| 128 | 128 |  512 | 512 |  2 |   806.9 | 237 | 106.5 |
| 128 | 128 |  768 | 512 |  2 |  1045.5 | 256 | 115.0 |
| 128 | 128 | 1536 | 512 |  2 |  1699.6 | 293 | 131.4 |
| 128 | 128 | 1664 | 512 |  2 |  1813.5 | 296 | 132.6 |

**DSv4 decode** (T=1):

| NH | topk | latency us |
|---:|---:|---:|
|  16 |  128 | 22.4 |
|  64 |  128 | 22.5 |
|  64 |  512 | 22.6 |
| 128 |  512 | 22.6 |
| 128 | 1024 | 24.6 |

**DSv3.2 / GLM-5.1** (656 B/slot inline, FP8 paged, `topk = 2048`):

| NH | T | latency us | KV GB/s | TFLOPs |
|---:|---:|---:|---:|---:|
|  16 |   1 |  22.7 |   59 |   3.2 |
|  16 |  64 | 137.9 |  624 |  33.1 |
|  16 | 256 | 352.3 |  976 |  51.8 |
|  64 |  16 | 111.5 |  193 |  40.9 |
|  64 | 256 | 454.7 |  756 | 160.6 |
| 128 |  16 | 163.9 |  131 |  55.7 |
| 128 | 256 | 811.0 |  424 | 180.1 |

## Testing

```bash
pytest tests/attention/test_sparse_mla_sm120.py
```

122 tests covering decode + prefill, single + dual cache, attention sink, top-k
length masking, and full DSv4 / DSv3.2 / GLM-5.1 shape coverage. Tests
`compute-sanitizer --tool racecheck` clean on representative shapes.

End-to-end validated by serving **GLM-5.1** and **DSv4-Flash** through these
kernels.

### CODE REVIEW

File: tests/attention/test_sparse_mla_sm120.py [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+import pytest
+import torch
+
+import flashinfer
+from flashinfer.utils import is_sm120a_supported
+
+pytestmark = pytest.mark.skipif(
+    not is_sm120a_supported(torch.device("cuda")),
+    reason="Sparse-MLA SM120 requires sm120a.",
+)
```
[the remark this answers is not in the document]
Comment (qsang-nv) [reply tests/attention/test_sparse_mla_sm120.py:45]: +1 to coderabbit, test should cover sm121 as pr describes.

**Comment by lucifer1004** [comment]
DSv4 dual-cache prefill results has been merged into PR body.

Summary: the final default path is 1.19-1.32x faster on the added DSv4 dual-cache prefill cases, with the largest gains on the non-standard large extra-topk cases.

Also addressed the SM121 coverage review comment by changing `tests/attention/test_sparse_mla_sm120.py` from `is_sm120a_supported(...)` to `is_sm12x_supported(...)`, so the suite is no longer skipped on supported SM121 runners.

**Comment by lukealonso** [comment]
LGTM

File: .gitignore [hunk 2]
```diff
@@ -28,6 +28,7 @@ flashinfer/cute_dsl/benchmark_gated_delta_rule.py
 
 # zed text editor
 .zed/
+
```
Comment (bkryu) [inline .gitignore]: Seems like an unnecessary diff?

File: flashinfer/jit/sparse_mla_sm120.py
Comment (bkryu) [inline flashinfer/jit/sparse_mla_sm120.py:1]: Can we put the contents of this file in [flashinfer/jit/mla.py](https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/jit/mla.py)?

File: flashinfer/mla/_core.py [hunk 3]
```diff
@@ -745,12 +1456,20 @@ def __init__(
             User-reserved buffer to back the ``kv_len_arr`` array, shape ``[batch_size]``,
             dtype ``int32``.  Only consulted when ``use_cuda_graph=True``.
         backend : str
-            The implementation backend, could be ``auto``/``fa2`` or ``fa3``. Defaults to ``auto``.
+            The implementation backend, could be ``auto``/``fa2``/``fa3``.
+            Defaults to ``auto``.
```
Comment (bkryu) [inline flashinfer/mla/_core.py]: Seems like an unnecessary change

File: flashinfer/mla/_core.py [hunk 4]
```diff
@@ -745,12 +1456,20 @@ def __init__(
...
             kernels will be generated by CUTLASS and only float_workspace_buffer is required and
             other arguments are ignored.
         """
+        if backend == "sparse-sm120":
+            raise ValueError(
+                "backend='sparse-sm120' is no longer supported on "
+                "BatchMLAPagedAttentionWrapper; use "
+                "trtllm_batch_decode_sparse_mla_dsv4 for DeepSeek V4 sparse MLA"
+            )
+
```
Comment (bkryu) [inline flashinfer/mla/_core.py]: Seems unnecessary. Probably a remnant from iterating with agents

File: flashinfer/mla/_sparse_mla_sm120.py [hunk 5]
```diff
@@ -0,0 +1,1346 @@
...
+    out_lse: torch.Tensor,
+    sm_scale: float,
+    *,
+    topk_length: Optional[torch.Tensor] = None,
+    attn_sink: Optional[torch.Tensor] = None,
+    extra_kv_cache: Optional[torch.Tensor] = None,
+    extra_indices: Optional[torch.Tensor] = None,
+    extra_topk_length: Optional[torch.Tensor] = None,
+    chunks_per_block: Optional[int] = None,
+) -> torch.Tensor:
```
Comment (bkryu) [inline flashinfer/mla/_sparse_mla_sm120.py]: As discussed previously, let's not expose a new API and rather put your kernel behind the existing `trtllm_...` API

File: flashinfer/mla/_core.py [hunk 6]
```diff
@@ -476,14 +1144,17 @@ def trtllm_batch_decode_sparse_mla_dsv4(
...
 
-    This API is for the TRTLLM-GEN DeepSeek V4 sparse MLA kernels where
-    ``headDimQk == headDimV == 512``. It supports BF16 and per-tensor FP8 E4M3
-    inputs with BF16 output. The SWA side is fixed to 128 indices per query.
-    Callers must provide a primary/compressed KV pool, the concatenated sparse
-    index matrix, and total sparse MLA top-k lengths for every query token.
+    Supports TRTLLM-GEN on SM100/SM103 and the SM120 sparse backend. For
+    ``backend="sparse"``, ``sparse_indices``/``swa_topk_lens`` describe the SWA
+    segment and ``extra_sparse_indices``/``extra_sparse_topk_lens`` describe the
+    optional compressed segment.
```
Comment (bkryu) [inline flashinfer/mla/_core.py]: I find the docstring here confusing. We are deleting information and adding confusion.

Can you try to re-write this while being mindful from a user's point of view? It should be clear how the API should be used for SM100/103, and SM120.

I would consider something like
```
    For SM 100/103, the API calls TRTLLM-GEN DeepSeek V4 sparse MLA kernels where
    ``headDimQk == headDimV == 512``. It supports BF16 and per-tensor FP8 E4M3
    inputs with BF16 output. The SWA side is fixed to 128 indices per query.
    Callers must provide a primary/compressed KV pool, the concatenated sparse
    index matrix, and total sparse MLA top-k lengths for every query token.

   For SM120/121, ....

```

File: flashinfer/mla/_core.py [hunk 7]
```diff
@@ -492,34 +1163,24 @@ def trtllm_batch_decode_sparse_mla_dsv4(
         or varlen query input ``[sum_q, num_heads, 512]`` when
         ``cum_seq_lens_q`` is provided. BF16 or FP8 E4M3.
     swa_kv_cache : torch.Tensor
-        SWA KV cache. HND layout is ``[num_pages, 1, page_size, 512]`` and NHD
-        layout is ``[num_pages, page_size, 1, 512]``. A 3D HND shorthand
-        ``[num_pages, page_size, 512]`` is also accepted.
+        SWA KV cache. TRTLLM-GEN uses head dim 512; SM120 sparse uses packed
+        uint8 head dim 584. Layout follows ``kv_layout``.
```
Comment (bkryu) [inline flashinfer/mla/_core.py:1171]: Again, we are losing information on the layout. Let's keep SM100/103 and SM120/121 information separately.

File: flashinfer/mla/_core.py [hunk 8]
```diff
@@ -492,34 +1163,24 @@ def trtllm_batch_decode_sparse_mla_dsv4(
...
-    compressed_kv_cache : torch.Tensor
-        Primary/compressed KV cache in the same layout and dtype as
-        ``swa_kv_cache``. For SWA-only calls, provide any valid primary pool.
-    sparse_topk_lens : torch.Tensor
+        TRTLLM-GEN combined sparse table, or the SM120 sparse SWA segment.
+    compressed_kv_cache : Optional[torch.Tensor]
+        Primary/compressed KV cache in the same backend layout as
+        ``swa_kv_cache``. Required by ``trtllm-gen`` and by SM120 ``sparse``
+        when ``extra_sparse_indices`` is provided.
+    sparse_topk_lens : Optional[torch.Tensor]
```
Comment (bkryu) [inline flashinfer/mla/_core.py:1180]: Ditto and in the docstring changes below.

File: flashinfer/mla/_core.py [hunk 9]
```diff
@@ -533,22 +1194,68 @@ def trtllm_batch_decode_sparse_mla_dsv4(
...
-        Maximum query length in the varlen batch. Required when
-        ``cum_seq_lens_q`` is provided so the wrapper does not need a
-        device-to-host synchronization to infer it.
+        Maximum query length in the varlen batch. Required with
+        ``cum_seq_lens_q``.
     enable_pdl : Optional[bool]
         Whether to enable Programmatic Dependent Launch.
+    backend : str
+        ``"auto"`` chooses ``"trtllm-gen"`` on SM100/SM103 and ``"sparse"``
+        on SM120.
```
Comment (bkryu) [inline flashinfer/mla/_core.py]: Do we need a `backend` parameter if an SM architecture maps to a unique kernel? The backend parameter might confuse the user.

File: flashinfer/mla/_core.py [hunk 10]
```diff
@@ -1671,13 +2391,16 @@ def trtllm_batch_decode_with_kv_cache_mla(
     qk_nope_head_dim: qk_nope_head_dim, must be 128 or 64
     kv_lora_rank: kv_lora_rank, must be 512 or 256
     qk_rope_head_dim: qk_rope_head_dim, must be 64
-    sparse_mla_top_k: sparse MLA top k, must be 0 for non-sparse MLA.
+    sparse_mla_top_k: sparse MLA top k for sparse page-table decode paths.
+        On SM100/SM103 this selects the TRTLLM-GEN sparse MLA path. On SM120,
+        ``backend="auto"`` selects the SM120 packed v32/GLM sparse backend.
```
Comment (bkryu) [inline flashinfer/mla/_core.py]: Docstring feels a bit weird.

File: flashinfer/mla/_sparse_mla_sm120.py
Comment (bkryu) [inline flashinfer/mla/_sparse_mla_sm120.py:1]: Can we move this file to `flashinfer/mla` instead of keeping it in `flashinfer`?

**Review by bkryu — COMMENTED** [review]
Thanks @lucifer1004 , left some comments on docstring and readability.

Can you update the PR description to keep it concise and make it very clear for frameworks that consider adopting the SM120 sparse MLA how to call the API?

File: flashinfer/mla/_core.py (hunk 4 above)
Comment (saltyminty) [reply flashinfer/mla/_core.py]: +1, backward compatibility is not needed here

File: flashinfer/mla/_core.py [hunk 11]
```diff
@@ -492,34 +1163,24 @@ def trtllm_batch_decode_sparse_mla_dsv4(
         or varlen query input ``[sum_q, num_heads, 512]`` when
         ``cum_seq_lens_q`` is provided. BF16 or FP8 E4M3.
```
Comment (saltyminty) [inline flashinfer/mla/_core.py]: If I'm not mistaken, the sm120 path only supports BF16?

File: flashinfer/mla/_sparse_mla_sm120.py
Comment (lucifer1004) [reply flashinfer/mla/_sparse_mla_sm120.py:1]: Moved.

File: flashinfer/mla/_core.py (hunk 11 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Corrected.

File: .gitignore (hunk 2 above)
Comment (lucifer1004) [reply .gitignore]: Fixed.

File: flashinfer/jit/sparse_mla_sm120.py
Comment (lucifer1004) [reply flashinfer/jit/sparse_mla_sm120.py:1]: Done.

File: flashinfer/mla/_core.py (hunk 3 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Removed.

File: flashinfer/mla/_core.py (hunk 4 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Removed.

File: flashinfer/mla/_sparse_mla_sm120.py (hunk 5 above)
Comment (lucifer1004) [reply flashinfer/mla/_sparse_mla_sm120.py]: Unregistered the function.

File: flashinfer/mla/_core.py (hunk 6 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Fixed.

File: flashinfer/mla/_core.py (hunk 7 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py:1171]: Fixed.

File: flashinfer/mla/_core.py (hunk 8 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py:1180]: Fixed.

File: flashinfer/mla/_core.py (hunk 9 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Removed.

File: flashinfer/mla/_core.py (hunk 10 above)
Comment (lucifer1004) [reply flashinfer/mla/_core.py]: Updated.

**Review by saltyminty — APPROVED** [review]

**Review by bkryu — APPROVED** [review]
